dev/TEAL_opcodes.md

```diff
@@ -665,376 +605,349 @@ At v2 it became allowed to branch to the end of the program exactly after the la
+- Availability: v2
 - Mode: Application
 
-params: Txn.Accounts offset (or, since v4, an account address that appears in Txn.Accounts or is Txn.Sender), state key, value.
+params: Txn.Accounts offset (or, since v4, an _available_ account address), state key, value.
 
 ## app_global_put
 
 - Opcode: 0x67
-- Pops: *... stack*, {[]byte A}, {any B}
-- Pushes: _None_
-- write key A and value B to global state of the current application
-- LogicSigVersion >= 2
+- Stack: ..., A: []byte, B &rarr; ...
+- write B to key A in the global state of the current application
+- Availability: v2
 - Mode: Application
 
 ## app_local_del
 
 - Opcode: 0x68
-- Pops: *... stack*, {any A}, {[]byte B}
-- Pushes: _None_
-- delete from account A local state key B of the current application
-- LogicSigVersion >= 2
+- Stack: ..., A, B: []byte &rarr; ...
+- delete key B from account A's local state of the current application
+- Availability: v2
 - Mode: Application
 
-params: Txn.Accounts offset (or, since v4, an account address that appears in Txn.Accounts or is Txn.Sender), state key.
+params: Txn.Accounts offset (or, since v4, an _available_ account address), state key.
 
 Deleting a key which is already absent has no effect on the application local state. (In particular, it does _not_ cause the program to fail.)
 
 ## app_global_del
 
 - Opcode: 0x69
-- Pops: *... stack*, []byte
-- Pushes: _None_
-- delete key A from a global state of the current application
-- LogicSigVersion >= 2
+- Stack: ..., A: []byte &rarr; ...
+- delete key A from the global state of the current application
+- Availability: v2
 - Mode: Application
 
 params: state key.
 
 Deleting a key which is already absent has no effect on the application global state. (In particular, it does _not_ cause the program to fail.)
 
-## asset_holding_get i
+## asset_holding_get f
 
 - Opcode: 0x70 {uint8 asset holding field index}
-- Pops: *... stack*, {any A}, {uint64 B}
-- Pushes: *... stack*, any, uint64
-- read from account A and asset B holding field X (imm arg) => {0 or 1 (top), value}
-- LogicSigVersion >= 2
+- Stack: ..., A, B: uint64 &rarr; ..., X: any, Y: uint64
+- X is field F from account A's holding of asset B. Y is 1 if A is opted into B, else 0
+- Availability: v2
 - Mode: Application
 
 `asset_holding_get` Fields:
 
 | Index | Name | Type | Notes |
-| --- | --- | --- | --- |
+| - | ------ | -- | --------- |
 | 0 | AssetBalance | uint64 | Amount of the asset unit held by this account |
 | 1 | AssetFrozen | uint64 | Is the asset frozen or not |
 
 
-params: Txn.Accounts offset (or, since v4, an account address that appears in Txn.Accounts or is Txn.Sender), asset id (or, since v4, a Txn.ForeignAssets offset). Return: did_exist flag (1 if the asset existed and 0 otherwise), value.
+params: Txn.Accounts offset (or, since v4, an _available_ address), asset id (or, since v4, a Txn.ForeignAssets offset). Return: did_exist flag (1 if the asset existed and 0 otherwise), value.
```

Comment:
I forgot: how do you differentiate between an asset ID and a `Txn.ForeignAssets` offset?

Reply:
If the number is less than the length of the static array, it is interpreted as an index into it.  While this seems annoyingly ill-defined for the first couple transactions of a new chain, it can't matter to testnet or mainnet, that did not have creatables at the start.  I have had a request to swap the meaning for the convenience of people with private chains that create things at the start.  Since we had not previously committed to a preference order, I've been considering it.  What do you think? (Under that definition, we would see if it is equal to a number in the array first, and only consider it an index after that. Need to think about how to deal with the extra cases that could make it _available_)

Reply:
From a purely aesthetic viewpoint, it makes more sense to see if the argument matches an ID of an available asset, and consider it an offset only if it does not. The questions are (a) is there a big difference in complexity between the implementations of these two orders, and (b) which is more convenient for applications to use.

For (a), I cannot imagine that there will be a big difference, since the implementation would anyway have to know the IDs of all the available assets.

For (b), an application that wants to use offsets would have to ensure that there are no available assets with that index, which sounds like a really hard thing to ensure in general. (The other direction is easier to deal with, if you want to use a very small ID you can look for it in the array and use its offset. But it is still quite annoying to have to do this.) Of course this only refers to very special cases on private networks.

So I guess the question is if we care about applications on private networks that want to use indexes and may have very small available asset-IDs. Perhaps just tell developers not to use offsets if they write applications that may have to work in such environments?